service/matching/tasklist/adaptive_scaler.go

```diff
@@ -0,0 +1,251 @@
+	ctx, cancel := context.WithCancel(context.Background())
+	return &adaptiveScalerImpl{
+		taskListID:     taskListID,
+		tlMgr:          tlMgr,
+		qpsTracker:     qpsTracker,
+		config:         config,
+		timeSource:     timeSource,
+		logger:         logger.WithTags(tag.ComponentTaskListAdaptiveScaler),
+		scope:          scope,
+		matchingClient: matchingClient,
+		taskListType:   getTaskListType(taskListID.GetType()),
+		ctx:            ctx,
+		cancel:         cancel,
+		overLoad:       false,
+		underLoad:      false,
+	}
+}
+
+func (a *adaptiveScalerImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&a.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+	a.wg.Add(1)
+	go a.runPeriodicLoop()
+}
+
+func (a *adaptiveScalerImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&a.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	a.cancel()
+	a.wg.Wait()
+}
+
+func (a *adaptiveScalerImpl) runPeriodicLoop() {
+	defer a.wg.Done()
+	timer := a.timeSource.NewTimer(a.config.AdaptiveScalerUpdateInterval())
+	defer timer.Stop()
+	for {
+		select {
+		case <-a.ctx.Done():
+			return
+		case <-timer.Chan():
+			a.logger.Debug("timer")
+			a.run()
+			timer.Reset(a.config.AdaptiveScalerUpdateInterval())
+		}
+	}
+}
+
+func (a *adaptiveScalerImpl) run() {
+	if !a.config.EnableAdaptiveScaler() || !a.config.EnableGetNumberOfPartitionsFromCache() {
+		return
+	}
+	qps := a.qpsTracker.QPS()
+	partitionConfig := a.tlMgr.TaskListPartitionConfig()
+	if partitionConfig == nil {
+		partitionConfig = &types.TaskListPartitionConfig{
+			NumReadPartitions:  1,
+			NumWritePartitions: 1,
+		}
+	}
+	// calculate the number of write partitions
+	upscaleThreshold := float64(a.config.PartitionUpscaleRPS())
+	downscaleThreshold := float64(a.config.PartitionDownscaleRPS())
+	if downscaleThreshold > upscaleThreshold {
+		downscaleThreshold = upscaleThreshold
+		a.logger.Warn("downscale threshold is larger than upscale threshold, use upscale threshold for downscale threshold instead")
+	}
+	numWritePartitions := partitionConfig.NumWritePartitions
+	numReadPartitions := partitionConfig.NumReadPartitions
+	a.scope.UpdateGauge(metrics.EstimatedAddTaskQPSGauge, qps)
+	if qps > upscaleThreshold {
+		if !a.overLoad {
+			a.overLoad = true
+			a.overLoadStartTime = a.timeSource.Now()
+		} else if a.timeSource.Now().Sub(a.overLoadStartTime) > a.config.PartitionUpscaleSustainedDuration() {
+			numWritePartitions = getNumberOfPartitions(partitionConfig.NumWritePartitions, qps, upscaleThreshold)
+			a.overLoad = false
+		}
+	} else {
+		a.overLoad = false
+	}
+	if qps < downscaleThreshold {
+		if !a.underLoad {
+			a.underLoad = true
+			a.underLoadStartTime = a.timeSource.Now()
+		} else if a.timeSource.Now().Sub(a.underLoadStartTime) > a.config.PartitionDownscaleSustainedDuration() {
+			numWritePartitions = getNumberOfPartitions(partitionConfig.NumWritePartitions, qps, upscaleThreshold) // NOTE: this has to be upscaleThreshold
```

Comment:
After thinking about this, I think to avoid fluctuation, the thresholds need to satisfy this inequality
```
downscale threshold <= upscale threshold * N / (N + 1) (for all N)
```
Let's take your (500, 600) thresholds as an example, if the global qps is 1801, then we would have 4 partitions. But if we have 4 partitions, the load of each partition will be around 450, which is less than 500, and we will need to downscale. When the downscale operation is triggered, we have to recalculate the number of partitions based on the qps of root partition. The estimation is around 450, so it could be 451 or 449. If the estimation is larger than 450, then we still have 4 partitions, so no change. But the estimation is 449, then downscale will be triggered.